quinn-proto/src/packet.rs

```diff
@@ -439,12 +440,11 @@ pub struct PartialEncode {
 }
 
 impl PartialEncode {
-    pub fn finish(
-        self,
-        buf: &mut [u8],
-        header_crypto: &RingHeaderCrypto,
-        crypto: Option<(u64, &Crypto)>,
-    ) {
+    pub fn finish<K, H>(self, buf: &mut [u8], header_crypto: &H, crypto: Option<(u64, &K)>)
+    where
+        K: crypto::Keys,
+        H: crypto::HeaderKeys,
```

Comment:
I don't think it can be done without requiring type annotations elsewhere. Also, conceptually, it's not clear that it makes sense, I think? The constraint for the type of `header_crypto` is just that it implements `HeaderKeys`; for this implementation, it's not required that it uses the same crypto library as `crypto`.